File: 23-merge-k-sorted-lists/python3/23-merge-k-sorted-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
      lists = [x for x in lists if x != None]
      if len(lists) == 0:
        return None
      ans = lists[0]
      rest = lists[1:]
      
      for r in rest:
        if r == None:
          next
        l = ans
        if l.val > r.val:
          l, r = r, l
        ans = l
        l = l.next
        w = ans
        while l != None and r != None:
          if l.val <= r.val:
            w.next = l
            l = l.next
          else:
            w.next = r
            r = r.next
          w = w.next
        if l == None and r != None:
          w.next = r
        if l != None and r == None:
          w.next = l

      return ans

File: 23-merge-k-sorted-lists/java/23-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
  public ListNode mergeKLists(ListNode[] lists) {
    var queue = new PriorityQueue<ListNode>(Comparator.comparing(v -> v.val));
    for (var n : lists) {
      if (n != null) {
        queue.add(n);
      }
    }

    ListNode result = queue.peek();
    ListNode node = null;
    while (!queue.isEmpty()) {
      var n = queue.poll();
      if (null != n.next) {
        queue.add(n.next);
      }
      if (null == node) {
        node = n;
      } else {
        node.next = n;
        node = node.next;
      }
      
      n.next = null;
    }

    return result;
  }
}